checker/jdk/nullness/src/java/util/concurrent/SynchronousQueue.java

```diff
@@ -1028,7 +1028,7 @@ public boolean remove(Object o) {
      * @param c the collection
      * @return {@code false} unless given collection is empty
      */
-    @Pure public boolean containsAll(Collection<?> c) {
+    @Pure public boolean containsAll(Collection<? extends @NonNull Object> c) {
```

Comment:
All the bulk methods on `SynchronousQueue`, plus the single-element `contains` and `remove`, are documented to suggest that null elements are permitted. (And an implementation would have to go out of its way to throw NPE for them.)

That said, I also grant that these bulk methods aren't documented to throw NPE for a null _collection_, and yet clearly doing so would be acceptable behavior (probably even _the only right behavior_ in the case of `containsAll`)....

Reply:
> All the bulk methods on SynchronousQueue, plus the single-element contains and remove, are documented to suggest that null elements are permitted.

Can you point me at that documentation?  I must be overlooking it, because I can't seem to find it.

Reply:
Sorry, "documented" in the sense that these methods do not include `@throws NullPointerException`, in contrast to methods like `put`. But as noted, bulk methods like `containsAll` _also_ don't include `@throws NullPointerException` for when the _collection itself_ is `null`, so we might not want to read too much into the lack of a `@throws` clause.

Reply:
Thanks for the clarification.  I don't want to read too much into that.  I'll leave them as is in the annotated JDK for now.